src/spaceapi_client.rs

```diff
@@ -1,84 +1,203 @@
 //extern crate spaceapi;
+use std::io::prelude::*;
 use rustc_serialize::json;
 use hyper::Client;
-use std::io::Read;
-use spaceapi::Optional::{Value,Absent};
+use spaceapi::Optional::{self, Value, Absent};
 use spaceapi::{Status};
 use spaceapi::sensors::{PeopleNowPresentSensor};
 
-struct SpaceApiClient {
+use std::time::Duration;
+use time::SteadyTime;
+use std::env;
+use std::fs::{self, File};
+use std::io;
+
+pub struct SpaceApiClient {
   last_fetch: i64,
+  webcams: Vec<String>,
+}
+
+impl SpaceApiClient {
+  pub fn new() -> SpaceApiClient {
+    let mut s = SpaceApiClient{
+      last_fetch: 0,
+      webcams: Vec::new(),
+    };
+    
+    s.fetch_webcams();
+    
+    s
+  }
+  
+  pub fn fetch_people_now_present(&self) -> ::std::result::Result<u64, String> {
+    let status = try!(fetch_status());
+    
+    extract_people_now_present(status)
+  }
+
+  pub fn get_webcams(&self) -> Vec<String> {
+    self.webcams.clone()
+  }
+
+  fn fetch_webcams(&mut self) {
+    if let Ok(status) = fetch_status() {
+      
+      if let Value(cams) = status.cam {
+        self.webcams = cams;
+      }
+    }
+  }
+  
+  pub fn get_tmp_path_for_webcam(&self, url :&String) -> Result<String,io::Error> {
+    let dir = env::temp_dir().join("coredump_bot").join("get_tmp_path_for_webcam");
+    let path = env::temp_dir().join("coredump_bot").join("get_tmp_path_for_webcam").join( self.basename(url) );
+    let path = path.as_path();
+    
+    
+    fs::create_dir_all(dir);
+    
+    let mut f = try!(File::create(&path));
+    
+    
+    let bin = try!(fetch_binary(url));
+    
+    
+    f.write_all(&bin);
+    f.sync_all();
+    
+    Ok::<String,io::Error>(format!("{}", path.to_str().unwrap()))
+  }
+  
+  pub fn basename(&self, path :&String) -> String {
+    let mut r = String::new();
+    
+    for c in path.chars() {
+      match c {
+        '/' => r.clear(),
+        _ => r.push( c ),
+      }
+    }
+    
+    r
+  }
```

Comment:
Ah, you want to use it on URLs. Neither the function name nor the parameter name (path) nor the documentation explains that :)

In that case, what about this implementation?

```rust
/// Return the last part of a path or URL.                                         
fn basename<'a>(path: &'a str) -> &'a str {
    match path.rfind('/') {
        Some(p) => &path[p+1..],
        None => &path,
    }
}

#[cfg(test)]
mod tests {
    use super::basename;

    #[test]
    fn path() {
        assert_eq!(basename("/usr/bin/bash"), "bash");
    }

    #[test]
    fn url() {
        assert_eq!(basename("http://coredump.ch/f/index.html"), "index.html");
    }

    #[test]
    fn plain() {
        assert_eq!(basename("foobar"), "foobar");
    }

    #[test]
    fn none() {
        assert_eq!(basename("/bin/"), "");
    }
}
```